impl_core/src/main/java/io/opencensus/implcore/metrics/DoubleGaugeImpl.java

```diff
@@ -0,0 +1,165 @@
+  }
+
+  @Override
+  public Point getDefaultTimeSeries() {
+    return registerTimeSeries(defaultLabelValues);
+  }
+
+  @Override
+  public void removeTimeSeries(List<LabelValue> labelValues) {
+    checkNotNull(labelValues, "labelValues should not be null.");
+    Utils.checkListElementNotNull(labelValues, "labelValues element should not be null.");
+
+    remove(labelValues);
+  }
+
+  @Override
+  public synchronized void clear() {
+    registeredTimeSeries.clear();
+  }
+
+  private synchronized void remove(List<LabelValue> labelValues) {
+    Map<List<LabelValue>, TimeSeriesProducer> registeredTimeSeriesCopy =
+        new HashMap<List<LabelValue>, TimeSeriesProducer>(registeredTimeSeries);
+    if (registeredTimeSeriesCopy.remove(labelValues) == null) {
+      // The element not present, no need to update the current map of time series.
+      return;
+    }
+    registeredTimeSeries = Collections.unmodifiableMap(registeredTimeSeriesCopy);
+  }
+
+  private synchronized <T> Point registerTimeSeries(List<LabelValue> labelValues) {
+    TimeSeriesProducer existingTimeSeries = registeredTimeSeries.get(labelValues);
+    if (existingTimeSeries != null) {
+      // Return a TimeSeries that are already registered.
+      return (PointImpl) existingTimeSeries;
+    }
+
+    TimeSeriesProducer newTimeSeries = new PointImpl(labelValues);
+    // Updating the map of time series happens under a lock to avoid multiple add operations
+    // to happen in the same time.
+    Map<List<LabelValue>, TimeSeriesProducer> registeredTimeSeriesCopy =
+        new HashMap<List<LabelValue>, TimeSeriesProducer>(registeredTimeSeries);
+    registeredTimeSeriesCopy.put(labelValues, newTimeSeries);
+    registeredTimeSeries = Collections.unmodifiableMap(registeredTimeSeriesCopy);
+
+    return (PointImpl) newTimeSeries;
+  }
+
+  @Override
+  public Metric getMetric(Clock clock) {
+    List<TimeSeries> timeSeriesList = new ArrayList<TimeSeries>(registeredTimeSeries.size());
+    for (Map.Entry<List<LabelValue>, TimeSeriesProducer> entry : registeredTimeSeries.entrySet()) {
+      timeSeriesList.add(entry.getValue().getTimeSeries(clock));
+    }
+    return Metric.create(metricDescriptor, timeSeriesList);
+  }
+
+  /** Implementation of {@link DoubleGauge.Point}. */
+  public static final class PointImpl extends Point implements TimeSeriesProducer {
+
+    // TODO(mayurkale): Consider to use DoubleAdder here, once we upgrade to Java8.
+    private final AtomicDouble value = new AtomicDouble(0);
+    private final List<LabelValue> labelValues;
+
+    PointImpl(List<LabelValue> labelValues) {
+      this.labelValues = labelValues;
+    }
+
+    @Override
+    public void inc() {
+      inc(1);
+    }
+
+    @Override
+    public void inc(double amt) {
+      value.addAndGet(amt);
+    }
+
+    @Override
+    public void dec() {
+      dec(1);
+    }
+
+    @Override
+    public void dec(double amt) {
+      value.addAndGet(-amt);
```

Comment:
This is hard to use and confusing in my opinion. What about we have inc, dec and add (which accepts negative values)? This will be more consistent with the AtomicDouble/AtomicLong APIs.

Reply:
/cc @fabxc any thoughts? I think inc(), inc(amt), dec(), dec(amt), set(amt) seems clear for me.